google/cloud/bigtable/data/_metrics/data_model.py

```diff
@@ -0,0 +1,451 @@
+    from google.cloud.bigtable.data._metrics.handlers._base import MetricsHandler
+
+
+LOGGER = logging.getLogger(__name__)
+
+# default values for zone and cluster data, if not captured
+DEFAULT_ZONE = "global"
+DEFAULT_CLUSTER_ID = "unspecified"
+
+# keys for parsing metadata blobs
+BIGTABLE_METADATA_KEY = "x-goog-ext-425905942-bin"
+SERVER_TIMING_METADATA_KEY = "server-timing"
+SERVER_TIMING_REGEX = re.compile(r".*gfet4t7;\s*dur=(\d+\.?\d*).*")
+
+INVALID_STATE_ERROR = "Invalid state for {}: {}"
+
+
+class OperationType(Enum):
+    """Enum for the type of operation being performed."""
+
+    READ_ROWS = "ReadRows"
+    SAMPLE_ROW_KEYS = "SampleRowKeys"
+    BULK_MUTATE_ROWS = "MutateRows"
+    MUTATE_ROW = "MutateRow"
+    CHECK_AND_MUTATE = "CheckAndMutateRow"
+    READ_MODIFY_WRITE = "ReadModifyWriteRow"
+
+
+class OperationState(Enum):
+    """Enum for the state of the active operation."""
+
+    CREATED = 0
+    ACTIVE_ATTEMPT = 1
+    BETWEEN_ATTEMPTS = 2
+    COMPLETED = 3
+
+
+@dataclass(frozen=True)
+class CompletedAttemptMetric:
+    """
+    An immutable dataclass representing the data associated with a
+    completed rpc attempt.
+
+    Operation-level fields (eg. type, cluster, zone) are stored on the
+    corresponding CompletedOperationMetric or ActiveOperationMetric object.
+    """
+
+    duration_ns: int
+    end_status: StatusCode
+    gfe_latency_ns: int | None = None
+    application_blocking_time_ns: int = 0
+    backoff_before_attempt_ns: int = 0
+    grpc_throttling_time_ns: int = 0
+
+
+@dataclass(frozen=True)
+class CompletedOperationMetric:
+    """
+    An immutable dataclass representing the data associated with a
+    completed rpc operation.
+
+    Attempt-level fields (eg. duration, latencies, etc) are stored on the
+    corresponding CompletedAttemptMetric object.
+    """
+
+    op_type: OperationType
+    uuid: str
+    duration_ns: int
+    completed_attempts: list[CompletedAttemptMetric]
+    final_status: StatusCode
+    cluster_id: str
+    zone: str
+    is_streaming: bool
+    first_response_latency_ns: int | None = None
+    flow_throttling_time_ns: int = 0
+
+
+@dataclass
+class ActiveAttemptMetric:
+    """
+    A dataclass representing the data associated with an rpc attempt that is
+    currently in progress. Fields are mutable and may be optional.
+    """
+
+    # keep monotonic timestamps for active attempts
+    start_time_ns: int = field(default_factory=lambda: time.monotonic_ns())
+    # the time taken by the backend, in nanoseconds. Taken from response header
+    gfe_latency_ns: int | None = None
+    # time waiting on user to process the response, in nanoseconds
+    # currently only relevant for ReadRows
+    application_blocking_time_ns: int = 0
+    # backoff time is added to application_blocking_time_ns
+    backoff_before_attempt_ns: int = 0
+    # time waiting on grpc channel, in nanoseconds
+    # TODO: capture grpc_throttling_time
+    grpc_throttling_time_ns: int = 0
```

Comment:
we realized that in java this metric also doesn't capture the time a request queued on the channel. So if it's hard to get it in python we can skip it.

Reply:
sounds good, maybe I'll remove this then